File: frontend/mobile/core/domain/src/commonMain/kotlin/edu/openwellness/mobile/core/domain/util/Result.kt
package edu.openwellness.mobile.core.domain.util

/**
 * Generic, typed success/failure wrapper used across every layer — data,
 * domain, presentation, and validation.
 *
 * NOTE: this shadows [kotlin.Result]. Always import this type explicitly
 * (`edu.openwellness.mobile.core.domain.util.Result`) at call sites.
 */
sealed interface Result<out D, out E : Error> {
    data class Success<out D>(val data: D) : Result<D, Nothing>

    // Fully-qualified bound so `Error` here is unambiguously the project marker
    // interface, never kotlin.Error.
    data class Error<out E : edu.openwellness.mobile.core.domain.util.Error>(
        val error: E,
    ) : Result<Nothing, E>
}

/** A [Result] that carries no success payload — only a typed error on failure. */
typealias EmptyResult<E> = Result<Unit, E>

inline fun <T, E : Error, R> Result<T, E>.map(
    map: (T) -> R,
): Result<R, E> {
    return when (this) {
        is Result.Error -> Result.Error(error)
        is Result.Success -> Result.Success(map(data))
    }
}

inline fun <T, E : Error> Result<T, E>.onSuccess(
    action: (T) -> Unit,
): Result<T, E> {
    return when (this) {
        is Result.Error -> this
        is Result.Success -> {
            action(data)
            this
        }
    }
}

inline fun <T, E : Error> Result<T, E>.onFailure(
    action: (E) -> Unit,
): Result<T, E> {
    return when (this) {
        is Result.Error -> {
            action(error)
            this
        }
        is Result.Success -> this
    }
}

fun <T, E : Error> Result<T, E>.asEmptyResult(): EmptyResult<E> {
    return map { }
}

File: frontend/dashboard/src/core/result.ts
// Generic typed success/failure wrapper. Mirrors mobile Result.kt
// (core/domain/util/Result.kt) — a discriminated union instead of a sealed
// interface, so `r.ok` narrows the value/error the same way `is Result.Success`
// does in Kotlin.

export type Result<T, E> = { ok: true; value: T } | { ok: false; error: E }

/** A Result that carries no success payload — only a typed error on failure. */
export type EmptyResult<E> = Result<void, E>

export function ok<T>(value: T): { ok: true; value: T } {
  return { ok: true, value }
}

/** Convenience for the void-payload case (mirrors `Result.Success(Unit)`). */
export function okVoid(): { ok: true; value: void } {
  return { ok: true, value: undefined }
}

export function err<E>(error: E): { ok: false; error: E } {
  return { ok: false, error }
}

export function isOk<T, E>(r: Result<T, E>): r is { ok: true; value: T } {
  return r.ok
}

export function map<T, U, E>(r: Result<T, E>, fn: (v: T) => U): Result<U, E> {
  return r.ok ? ok(fn(r.value)) : r
}

export function mapErr<T, E, F>(
  r: Result<T, E>,
  fn: (e: E) => F,
): Result<T, F> {
  return r.ok ? r : err(fn(r.error))
}
